File: solution2.py
# https://leetcode.com/problems/delete-and-earn

# // Time Complexity : o(len(A)^2)
# // Space Complexity : o(len(A)^2)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this : None
#
#
# // Your code here along with comments explaining your approach

class Solution:
    def minFallingPathSum(self, A: List[List[int]]) -> int:
        _max = 9999

        dp = [[_max for i in range(len(A))] for j in range(len(A))]
        for i in range(len(A)):
            dp[0][i] = A[0][i]

        for i in range(1,len(A)):
            for j in range(len(A)):
                if j == 0:
                    dp[i][j] = A[i][j] + min(dp[i - 1][0], dp[i - 1][1])
                elif j == len(A) - 1:
                    dp[i][j] = A[i][j] + min(dp[i - 1][-1], dp[i - 1][-2])
                else:
                    dp[i][j] = A[i][j] + min(dp[i - 1][j - 1], dp[i - 1][j], dp[i - 1][j + 1])
        return min(dp[len(A) - 1])

File: Solution2.java
// Leetcode Problem :  https://leetcode.com/problems/minimum-falling-path-sum/description/
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*
 * I start traversing from the first row of the matrix, I obtain the minimum from elements that are directly above or diagonally left/right in the 
 * previous row and add the value of current element to it. After doing this, I obtain the minimum of the last row and return that
 * as the answer.
 */
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int rows = matrix.length;
        int columns = matrix[0].length;
        for(int i=1;i<rows;i++)
        {
            for(int j=0;j<columns;j++)
            {
                if(j==0)
                {
                    matrix[i][j]+=Math.min(matrix[i-1][j],matrix[i-1][j+1]);
                }
                else if(j==columns-1)
                {
                    matrix[i][j]+=Math.min(matrix[i-1][j],matrix[i-1][j-1]);
                }
                else
                {
                    matrix[i][j]+=Math.min(matrix[i-1][j],Math.min(matrix[i-1][j+1],matrix[i-1][j-1]));
                }
            }
        }


        int min = Integer.MAX_VALUE;
        for(int i=0;i<columns;i++)
        {
            min=Math.min(min,matrix[rows-1][i]);
        }
        return min;
    }
}